pulser-core/pulser/backend/default_observables.py

```diff
@@ -0,0 +1,369 @@
+                for j in range(state.n_qudits)
+            ]
+            for i in range(state.n_qudits)
+        ]
+
+
+class QubitDensity(Observable):  # TODO: Rename
+    """Stores the population of an eigenstate on each qudit.
+
+    Args:
+        evaluation_times: The relative times at which to compute the
+            correlation matrix. If left as `None`, uses the
+            `default_evaluation_times` of the backend's `EmulationConfig`.
+        one_state: The eigenstate to measure the population of. Can be left
+            undefined if the state's eigenstates form a known eigenbasis with
+            a defined "one state".
+    """
+
+    def __init__(
+        self,
+        *,
+        evaluation_times: Sequence[float] | None = None,
+        one_state: Eigenstate | None = None,
+    ):
+        """Initializes the observable."""
+        super().__init__(evaluation_times=evaluation_times)
+        # TODO: Checks
+        self.one_state = one_state
+
+    def name(self) -> str:
+        """Returns the name of the observable."""
+        return "qubit_density"
+
+    def apply(
+        self, *, state: State, hamiltonian: Operator, **kwargs: Any
+    ) -> list:
+        """Calculates the observable to store in the Results."""
+        return [
+            CorrelationMatrix._get_number_operator(
+                frozenset((i,)),
+                state.n_qudits,
+                state.eigenstates,
+                self.one_state or state.infer_one_state(),
+                type(hamiltonian),
+            ).expect(state)
+            for i in range(state.n_qudits)
+        ]
+
+
+class Energy(Observable):
+    """Stores the energy of the system at the evaluation times.
+
+    Args:
+        evaluation_times: The relative times at which to compute the energy.
+            If left as `None`, uses the `default_evaluation_times` of the
+            backend's `EmulationConfig`.
+    """
+
+    def name(self) -> str:
+        """Returns the name of the observable."""
+        return "energy"
+
+    def apply(
+        self, *, state: State, hamiltonian: Operator, **kwargs: Any
+    ) -> Any:
+        """Calculates the observable to store in the Results."""
+        return hamiltonian.expect(state)
+
+
+class EnergyVariance(Observable):
+    """Stores the varaiance of the Hamiltonian at the evaluation times.
+
+    Args:
+        evaluation_times: The relative times at which to compute the variance.
+            If left as `None`, uses the `default_evaluation_times` of the
+            backend's `EmulationConfig`.
+    """
+
+    def name(self) -> str:
+        """Returns the name of the observable."""
+        return "energy_variance"
+
+    def apply(
+        self, *, state: State, hamiltonian: Operator, **kwargs: Any
+    ) -> Any:
+        """Calculates the observable to store in the Results."""
+        # This works for state vectors and density matrices and avoids
+        # squaring the hamiltonian
```

Comment:
last week I found the following implementation to be significantly faster and more memory efficient

```
h_squared = H @ H
        return h_squared.expect(state).real - H.expect(state).real ** 2
```

Reply:
Interesting, I thought that matrix-matrix multiplication would always be more resource intensive, so I tried to avoid it.  Just goes to show how little I know about matrix product states, I guess